tools/clang/test/HLSLFileCheck/samples/d3d11/ComputeShaderSort11.hlsl

```diff
@@ -6,10 +6,10 @@
 // CHECK: addrspace(3)
 // CHECK: barrier
 // CHECK: addrspace(3)
-// CHECK: barrier
 // CHECK: addrspace(3)
 // CHECK: barrier
 // CHECK: addrspace(3)
+// CHECK: barrier
```

Comment:
Makes sense. In this case it's just different code layout, two blocks have moved higher in the block list, nothing else. This does change the order of addrspace/barrier occurrences in the code, but then again there's many more occurrences of addrspace(3) than what's being checked here, so not sure how useful these checks are to begin with. There is a third barrier in the code even before my change, so maybe something that has changed since introduction of the test? Possibly worth doing this instead to prevent future regressions:
```
// CHECK: barrier
// CHECK: barrier
// CHECK: barrier
// CHECK-NOT: barrier
```

Here's the before/after IR:
```
define void @main() {
entry:
  %Data_UAV_structbuf = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 1, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %CB_cbuffer = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 2, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %0 = call i32 @dx.op.threadId.i32(i32 93, i32 0)  ; ThreadId(component)
  %1 = call i32 @dx.op.flattenedThreadIdInGroup.i32(i32 96)  ; FlattenedThreadIdInGroup()
  %2 = call %dx.types.ResRet.i32 @dx.op.bufferLoad.i32(i32 68, %dx.types.Handle %Data_UAV_structbuf, i32 %0, i32 0)  ; BufferLoad(srv,index,wot)
  %3 = extractvalue %dx.types.ResRet.i32 %2, 0
  %arrayidx = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %1
  store i32 %3, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  %4 = call %dx.types.CBufRet.i32 @dx.op.cbufferLoadLegacy.i32(i32 59, %dx.types.Handle %CB_cbuffer, i32 0)  ; CBufferLoadLegacy(handle,regIndex)
  %5 = extractvalue %dx.types.CBufRet.i32 %4, 0
  %shr = lshr i32 %5, 1
  %cmp.16 = icmp eq i32 %shr, 0
  br i1 %cmp.16, label %for.end, label %for.body.preheader

for.body.preheader:                               ; preds = %entry
  br label %for.body

for.body:                                         ; preds = %for.body, %for.body.preheader
  %j.0 = phi i32 [ %shr13, %for.body ], [ %shr, %for.body.preheader ]
  %neg = xor i32 %j.0, -1
  %and = and i32 %1, %neg
  %arrayidx2 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %and
  %6 = load i32, i32 addrspace(3)* %arrayidx2, align 4, !tbaa !25
  %or = or i32 %j.0, %1
  %arrayidx3 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %or
  %7 = load i32, i32 addrspace(3)* %arrayidx3, align 4, !tbaa !25
  %cmp4 = icmp ule i32 %6, %7
  %8 = call %dx.types.CBufRet.i32 @dx.op.cbufferLoadLegacy.i32(i32 59, %dx.types.Handle %CB_cbuffer, i32 0)  ; CBufferLoadLegacy(handle,regIndex)
  %9 = extractvalue %dx.types.CBufRet.i32 %8, 1
  %and6 = and i32 %9, %0
  %tobool7 = icmp ne i32 %and6, 0
  %cmp8tmp = xor i1 %cmp4, %tobool7
  %xor = xor i32 %j.0, %1
  %arrayidx10 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %xor
  %10 = load i32, i32 addrspace(3)* %arrayidx10, align 4, !tbaa !25
  %11 = load i32, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  %12 = select i1 %cmp8tmp, i32 %11, i32 %10
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  store i32 %12, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  %shr13 = lshr i32 %j.0, 1
  %cmp = icmp eq i32 %shr13, 0
  br i1 %cmp, label %for.end.loopexit, label %for.body

for.end.loopexit:                                 ; preds = %for.body
  br label %for.end

for.end:                                          ; preds = %for.end.loopexit, %entry
  %13 = load i32, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.bufferStore.i32(i32 69, %dx.types.Handle %Data_UAV_structbuf, i32 %0, i32 0, i32 %13, i32 undef, i32 undef, i32 undef, i8 1)  ; BufferStore(uav,coord0,coord1,value0,value1,value2,value3,mask)
  ret void
}
```

```
define void @main() {
entry:
  %Data_UAV_structbuf = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 1, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %CB_cbuffer = call %dx.types.Handle @dx.op.createHandle(i32 57, i8 2, i32 0, i32 0, i1 false)  ; CreateHandle(resourceClass,rangeId,index,nonUniformIndex)
  %0 = call i32 @dx.op.threadId.i32(i32 93, i32 0)  ; ThreadId(component)
  %1 = call i32 @dx.op.flattenedThreadIdInGroup.i32(i32 96)  ; FlattenedThreadIdInGroup()
  %2 = call %dx.types.ResRet.i32 @dx.op.bufferLoad.i32(i32 68, %dx.types.Handle %Data_UAV_structbuf, i32 %0, i32 0)  ; BufferLoad(srv,index,wot)
  %3 = extractvalue %dx.types.ResRet.i32 %2, 0
  %arrayidx = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %1
  store i32 %3, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  %4 = call %dx.types.CBufRet.i32 @dx.op.cbufferLoadLegacy.i32(i32 59, %dx.types.Handle %CB_cbuffer, i32 0)  ; CBufferLoadLegacy(handle,regIndex)
  %5 = extractvalue %dx.types.CBufRet.i32 %4, 0
  %shr = lshr i32 %5, 1
  %cmp.16 = icmp eq i32 %shr, 0
  br i1 %cmp.16, label %for.cond.cleanup, label %for.body.preheader

for.body.preheader:                               ; preds = %entry
  br label %for.body

for.cond.cleanup.loopexit:                        ; preds = %for.body
  br label %for.cond.cleanup

for.cond.cleanup:                                 ; preds = %for.cond.cleanup.loopexit, %entry
  %6 = load i32, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.bufferStore.i32(i32 69, %dx.types.Handle %Data_UAV_structbuf, i32 %0, i32 0, i32 %6, i32 undef, i32 undef, i32 undef, i8 1)  ; BufferStore(uav,coord0,coord1,value0,value1,value2,value3,mask)
  ret void

for.body:                                         ; preds = %for.body, %for.body.preheader
  %j.0 = phi i32 [ %shr13, %for.body ], [ %shr, %for.body.preheader ]
  %neg = xor i32 %j.0, -1
  %and = and i32 %1, %neg
  %arrayidx2 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %and
  %7 = load i32, i32 addrspace(3)* %arrayidx2, align 4, !tbaa !25
  %or = or i32 %j.0, %1
  %arrayidx3 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %or
  %8 = load i32, i32 addrspace(3)* %arrayidx3, align 4, !tbaa !25
  %cmp4 = icmp ule i32 %7, %8
  %9 = call %dx.types.CBufRet.i32 @dx.op.cbufferLoadLegacy.i32(i32 59, %dx.types.Handle %CB_cbuffer, i32 0)  ; CBufferLoadLegacy(handle,regIndex)
  %10 = extractvalue %dx.types.CBufRet.i32 %9, 1
  %and6 = and i32 %10, %0
  %tobool7 = icmp ne i32 %and6, 0
  %cmp8tmp = xor i1 %cmp4, %tobool7
  %xor = xor i32 %j.0, %1
  %arrayidx10 = getelementptr [512 x i32], [512 x i32] addrspace(3)* @"\01?shared_data@@3PAIA", i32 0, i32 %xor
  %11 = load i32, i32 addrspace(3)* %arrayidx10, align 4, !tbaa !25
  %12 = load i32, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  %13 = select i1 %cmp8tmp, i32 %12, i32 %11
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  store i32 %13, i32 addrspace(3)* %arrayidx, align 4, !tbaa !25
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  %shr13 = lshr i32 %j.0, 1
  %cmp = icmp eq i32 %shr13, 0
  br i1 %cmp, label %for.cond.cleanup.loopexit, label %for.body
}
```

Reply:
Is there anything that can be done to fix the ordering of loopexit and cleanup before the loop body with this change?

Reply:
Ok, I think we can live with this for now then.